### ISSUE TITLE
The language server crashes in v1.21.x on macOS Big Sur or older

### ISSUE THREAD

**Issue by coshz** [microsoft/vscode-cpptools#12511]
### Environment

- OS and Version: macOS 10.15.7
- VS Code Version: 1.91.1
- C/C++ Extension Version: v1.21.2, v1.21.3

### Bug Summary and Steps to Reproduce

Bug Summary: The language server crashes.
Steps to reproduce: Install v1.21.2 or v1.21.3 and then open a c++ source file.
Expected behavior: No crash.

### Configuration and Logs

1. configurations in `c_cpp_properties.json`: default;
2. logs from running `c/c++: log Diagnostics`: no response;
3. logs from the language server logging:
```
[Error - 6:25:10 PM] Server process exited with signal SIGSEGV.
[Error - 6:25:10 PM] Server initialization failed.
  Message: Pending response rejected since connection got disposed
  Code: -32097 
[Error - 6:25:10 PM] The language server crashed. Restarting...
[Error - 6:25:10 PM] cpptools client: couldn't create connection to server.
  Message: Pending response rejected since connection got disposed
  Code: -32097 
```
and
```
[Error - 6:25:11 PM] Server initialization failed.
  Message: Pending response rejected since connection got disposed
  Code: -32097 
[Error - 6:25:11 PM] The language server crashed 5 times in the last 3 minutes. It will not be restarted.
[Error - 6:25:11 PM] cpptools client: couldn't create connection to server.
  Message: Pending response rejected since connection got disposed
  Code: -32097 
[Error - 6:25:11 PM] Server process exited with signal SIGSEGV.
```

**Comment by sean-mcmanus** [comment]
@coshz I would guess it doesn't repro for you with 1.20.5 -- can you confirm? Does a "C/C++ Crash Call Stacks" output window pane appear? If not, are you able to provide a crash call stack by attaching to the cpptools process before opening the file? Or are you able to enable crash call stack logging with your macOS?

**Comment by coshz** [comment]
> @coshz I would guess it doesn't repro for you with 1.20.5 -- can you confirm? Does a "C/C++ Crash Call Stacks" output window pane appear? If not, are you able to provide a crash call stack by attaching to the cpptools process before opening the file? Or are you able to enable crash call stack logging with your macOS?

Yes, v1.20.5 works. There is no such "C/C++ Crash Call Stacks" output pane; about "crash call stack", do you mean the console log of "Help > Toggle developer tools"? If yes, please see:
[vscode-app-1721974864912.log](https://github.com/user-attachments/files/16387425/vscode-app-1721974864912.log)

**Comment by sean-mcmanus** [comment]
@coshz The "crash call stacks" logging looks like 
![image](https://github.com/user-attachments/assets/fe4ef34c-a511-44d4-aac3-3b13c809ad5b)

but crash call stacks are not available for certain scenarios.

We haven't been able to get a repro yet. One guess is that it might require a particular OS version or SSH extension version.

**Comment by coshz** [comment]
> @coshz The "crash call stacks" logging looks like ![image](https://private-user-images.githubusercontent.com/19859882/352629161-fe4ef34c-a511-44d4-aac3-3b13c809ad5b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIwNjQ5MzEsIm5iZiI6MTcyMjA2NDYzMSwicGF0aCI6Ii8xOTg1OTg4Mi8zNTI2MjkxNjEtZmU0ZWYzNGMtYTUxMS00NGQ0LWFhYzMtM2IxM2M4MDlhZDViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI3VDA3MTcxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNiZTc4ZDFiMDgwODNiMzVjNTMyYmU4MzA1ODU3MmJhNzgyM2E5MjAyN2YxMjBlMjRlOTI0NGI2ODEyZTg2YjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.W0pLqiWPQ-W9qBJjHKQwzKLIMKaK2e8MgIxCSBFTNGA)
> 
> but crash call stacks are not available for certain scenarios.
> 
> We haven't been able to get a repro yet. One guess is that it might require a particular OS version or SSH extension version.

My vscode doesn't have the "C/C++ Crash Call Stacks" output panel. Attaching debugger to cpptools (reference [this](https://github.com/microsoft/vscode-cpptools/wiki/Attaching-debugger-to-cpptools-or-cpptools%E2%80%90srv)) seems infeasible since cpptools doesn't appear in the pid picker. By the way, this issue occurs when vscode starts, and opening a C++ file is unnecessary.
The following is the result of `lldb cpptools` (which might be helpful?):
```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x0000000100d9a28b cpptools`__cxx_global_var_init.13 + 11
    frame #1: 0x0000000101c1f353 dyld`ImageLoaderMachO::doModInitFunctions(ImageLoader::LinkContext const&) + 535
    frame #2: 0x0000000101c1f75e dyld`ImageLoaderMachO::doInitialization(ImageLoader::LinkContext const&) + 40
    frame #3: 0x0000000101c1a17b dyld`ImageLoader::recursiveInitialization(ImageLoader::LinkContext const&, unsigned int, char const*, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 493
    frame #4: 0x0000000101c18234 dyld`ImageLoader::processInitializers(ImageLoader::LinkContext const&, unsigned int, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 188
    frame #5: 0x0000000101c182d4 dyld`ImageLoader::runInitializers(ImageLoader::LinkContext const&, ImageLoader::InitializerTimingList&) + 82
    frame #6: 0x0000000101c066a8 dyld`dyld::initializeMainExecutable() + 199
    frame #7: 0x0000000101c0bbf7 dyld`dyld::_main(macho_header const*, unsigned long, int, char const**, char const**, char const**, unsigned long*) + 6728
    frame #8: 0x0000000101c05227 dyld`dyldbootstrap::start(dyld3::MachOLoaded const*, int, char const**, dyld3::MachOLoaded const*, unsigned long*) + 453
    frame #9: 0x0000000101c05025 dyld`_dyld_start + 37
```

**Comment by sean-mcmanus** [comment]
@coshz We believe this only repros on MacOS versions Big Sur and Catalina. It's possible it could be some bug in their OS and/or the SDKs we build with targetting those older OS's, i.e. we have our SDK set to target Mac Sierra and later, but if the SDK has a bug it could be causing some library reference that isn't available on the older OS's.

**Comment by socay1** [comment]
Yes, upgrade OS to Sonoma 14.6，the issue is corrected.
Great. @sean-mcmanus

**Comment by sean-mcmanus** [comment]
@coshz @socay1 It was caused by a bug in the XCode 15.0.1 SDK that they fixed in 15.1:

https://developer.apple.com/documentation/xcode-release-notes/xcode-15_1-release-notes#Resolved-Issues

> Fixed: Binaries using symbols with a weak definition crash at runtime on iOS 14/macOS 12 or older. This impacts primarily C++ > projects due to their extensive use of weak symbols. (114813650) (FB13097713)

We'll release a fix in our next update: https://github.com/microsoft/vscode-cpptools/milestone/176

**Comment by sean-mcmanus** [comment]
@coshz Fixed with https://github.com/microsoft/vscode-cpptools/releases/tag/v1.21.5
